### ISSUE TITLE
How to import 16 or 32 bit images properly?

### ISSUE THREAD

**Issue by cakwok** [ultralytics/ultralytics#1107]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I have tried to import 16 bit .tiff images into Yolo8, updated couple places from imread with imread_anydepth | imread_anycolor, have seen it were successfully imported by cv2 as uint16 at base.py with the following function, but I still see the images have become 8bit at augment.py class Albumentation.

def load_image(self, i):
    # Loads 1 image from dataset index 'i', returns (im, resized hw)
    im, f, fn = self.ims[i], self.im_files[i], self.npy_files[i]
    if im is None:  # not cached in RAM
        if fn.exists():  # load npy
            im = np.load(fn)
        else:  # read image
            im = cv2.imread(f, cv2.IMREAD_ANYDEPTH | cv2.IMREAD_ANYCOLOR)        <--- changes

Could you share with me the proper way to load 16 bit or 32 bit images? Thanks.

**Comment by glenn-jocher** [comment]
You can try using the flag cv2.IMREAD_UNCHANGED when reading the 16 or 32-bit TIFF images with OpenCV's imread() function. This flag ensures that the image is loaded with the same bit depth and number of channels as the original image. For example:

```python
im = cv2.imread(f, cv2.IMREAD_UNCHANGED)
```

Alternatively, you can use the Python Imaging Library (PIL) to read 16 or 32-bit TIFF images and convert them to NumPy arrays. You can then use these arrays as input to YOLOv8. Here's an example:

```python
from PIL import Image
import numpy as np

im = Image.open(f)
im = np.array(im)
```

Make sure to test these methods with your specific image format and ensure that the bit depth and channels are correctly preserved.

**Comment by cakwok** [comment]
Dear Glenn,

Thanks a lot for your reply!  Yes I understand the syntax requirement when I am reading a 16 bit and 32 bit image.  However, I would like to know from the end-to-end system perspective, how to ensure a read image is preserved the precision points all the way in the training.  I would like to know where shall I update the normalization rules accordingly, plus how to turn off any process which would return me a 8 bit image such as RandomHSV at data augmentation which I have encountered.

Could you help to provide the information?  Thanks.

**Comment by pderrenger** [comment]
@cakwok absolutely, I understand your concern. When dealing with 16 or 32 bit images in an end-to-end training pipeline with YOLOv8, it's important to ensure that the precision is maintained throughout the process.

One crucial aspect is to adjust the normalization process. Typically, in the data augmentation stage, some transformations may inadvertently convert the image to 8-bit format. To address this, you can customize the data augmentation pipeline to preserve the bit depth by creating a custom Albumentations transformation that explicitly retains the original bit depth of the input images.

Additionally, within the model definition and training script, you may need to adjust the normalization and data processing steps to handle 16 or 32 bit images appropriately. This can involve modifying the normalization parameters and ensuring that the model input and output layers are compatible with the desired precision.

By customizing the data augmentation and normalization steps in the training script, you can maintain the precision of your input images throughout the training process. Let me know if you need further assistance with implementing these changes!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR #20223, which introduces comprehensive multispectral image support throughout the Ultralytics ecosystem.

Key changes include:
- Full support for multi-channel datasets. You can now specify the number of channels in your dataset config, enabling seamless training, validation, prediction, and export with images that have more than 3 channels (such as 10-channel multispectral data).
- All YOLO model trainers, augmentations, and preprocessing steps now dynamically adjust to the dataset’s channel count.
- Model export and inference pipelines have been updated to handle and record the correct number of input channels, with metadata included for exported models.
- Visualization functions now gracefully display multispectral images by using only the first three channels.
- Additional tests ensure reliable multi-channel handling across different workflows.

Special thanks to @Laughing-q for the main implementation and to the contributors @glenn-jocher and @Y-T-G for their support.

To test if this resolves your issue, you can install the latest update:
```
pip install -U ultralytics>=8.3.12
```
Alternatively, you can install directly from the main branch for the most recent changes:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Or, if you prefer, you can wait for the next official PyPI release.

Please let us know if the changes in this update resolve your issue. Thank you for reporting this, and feel free to share any further feedback if you encounter any other problems! 🙏

**Comment by glenn-jocher** [comment]
@cakwok I see your concern about preserving bit depth throughout the entire pipeline. The good news is that a recent update (v8.3.112+) has implemented comprehensive multispectral image support that should resolve your issue.

For TIFF images specifically, the Ultralytics codebase now includes specialized handling via `cv2.imdecodemulti` that preserves high bit depth. Looking at the patches module, there's specific handling for TIFF files:

```python
def imread(filename: str, flags: int = cv2.IMREAD_COLOR):
    file_bytes = np.fromfile(filename, np.uint8)
    if filename.endswith((".tiff", ".tif")):
        success, frames = cv2.imdecodemulti(file_bytes, cv2.IMREAD_UNCHANGED)
        if success:
            return frames[0] if len(frames) == 1 and frames[0].ndim == 3 else np.stack(frames, axis=2)
```

To ensure your workflow preserves bit depth, make sure to:
1. Update to the latest version (`pip install -U ultralytics>=8.3.112`)
2. Specify the number of channels in your dataset YAML file with `channels: X` (where X is your channel count)
3. Prepare your TIFF images in `(channel, height, width)` order when saving them

The pipeline should now automatically adjust augmentations and normalization to preserve your image precision throughout training.
